### DISCUSSION TITLE
Error building ICU4X on a no_std platform

### DISCUSSION THREAD

**Discussion by ghost** [unicode-org/icu4x#5115] [General]
I'm following [Using ICU4X from C++](https://unicode-org.github.io/icu4x/cppdoc/index.html) to compile a c++ static library. We will be building it on a no_std platform, more specifically `thumbv7em-none-eabi`, so I used the command listed under [Embedded Platforms](https://unicode-org.github.io/icu4x/cppdoc/index.html#autotoc_md4) to compile. However, I'm getting errors like
```
   Compiling icu_capi v1.5.0
error[E0152]: found duplicate lang item `panic_impl`
  --> /usr/local/google/home/jiyuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/icu_capi-1.5.0/src/lib.rs:51:1
   |
51 | / fn panic(_info: &core::panic::PanicInfo) -> ! {
52 | |     loop {}
53 | | }
   | |_^
   |
   = note: the lang item is first defined in crate `std` (which `serde` depends on)
   = note: first definition in `std` loaded from /usr/local/google/home/jiyuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-d5189b81a4fa4d36.rlib
   = note: second definition in the local crate (`icu_capi`)

For more information about this error, try `rustc --explain E0152`.
error: could not compile `icu_capi` (lib) due to 1 previous error
```
if the target is not specified (I guess it's using default `stable-x86_64-unknown-linux-gnu`) and much more errors if the target is set to `thumbv7em-none-eabi`.

It seems like the errors come from the dependency "serde", which is dependent on "std". So I tried specifying `serde = { version = "1.0", default-features = false }` in `Cargo.toml`, but that doesn't help. How would you suggest to compile ICU4X on a no_std platform?

**Comment by Manishearth** [comment]
Looks like some crate is including serde with non-default features.

**Comment by Manishearth** [comment]
I'm unable to reproduce this, could you share the `cargo` command and the `Cargo.toml` you're using?

**Reply by Manishearth** [reply]
Basically you need to build `icu_capi` with `--no-default-features` or `default-features = false` depending on how you're using it.

**Reply by Manishearth** [reply]
Ah, I'm able to repro now

**Reply by Manishearth** [reply]
So the first problem is that you're not invoking it with the right `--target` flag

Try

`cargo rustc --release -p icu_capi --crate-type staticlib --features icu_capi/default_components,icu_capi/buffer_provider,icu_capi/looping_panic_handler,icu_capi/libc_alloc  --target thumbv7em-none-eabi`


However, that's also broken. 

For now if you run `cargo rustc --release -p icu_capi --crate-type staticlib --features icu_capi/default_components,icu_capi/buffer_provider,icu_capi/looping_panic_handler,icu_capi/libc_alloc  --no-default-features --target thumbv7em-none-eabi` from ICU4X's tree with https://github.com/unicode-org/icu4x/pull/5119 applied, it should _mostly_ work?

**Reply by Manishearth** [reply]
@sffc So our tutorial's way of building icu_capi is _weird_, we use `cargo rustc -p icu_capi` with a fake crate

With the tutorial steps, and the following command, I still get a problem:

```
$ cargo rustc --release -p icu_capi --crate-type staticlib --features icu_capi/default_components,icu_capi/buffer_provider,icu_capi/looping_panic_handler,icu_capi/libc_alloc  --target thumbv7em-none-eabi 2>&1 | less
warning: unused manifest key: resolver
   Compiling serde v1.0.203
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv7em-none-eabi` target may not support the standard library
  = note: `std` is required by `serde` because it does not declare `#![no_std]`
```



However, `cargo tree -e features -p icu_capi` does not show up `serde` anywhere except under `diplomat`, which is a proc macro.

**Reply by Manishearth** [reply]
Got it, we need `resolver = 2`. Preparing a docs fix now

**Reply by Manishearth** [reply]
Fixing the docs: https://github.com/unicode-org/icu4x/pull/5122

I'm also going to fix diplomat: https://github.com/rust-diplomat/diplomat/pull/510

There should be patch releases for these things available soon.

**Comment by Manishearth** [comment]
I found one bug in the libc_alloc feature, though: https://github.com/unicode-org/icu4x/pull/5119

**Comment by Manishearth** [comment]
The docs are being updated in https://github.com/unicode-org/icu4x/pull/5122

`diplomat_core` was released at 0.8.1 with this patch: https://github.com/rust-diplomat/diplomat/pull/510 . This will make it work with the older docs, but it should already work with the new docs, barring the `icu_capi` bug below.

I also discovered a bug in `icu_capi` that prevents this build from working, this has been fixed in https://github.com/unicode-org/icu4x/pull/5123 and patch-released as `icu_capi` 1.5.1.

Thanks for reporting this to us!

**Comment by ghost** [comment]
@Manishearth Thank you so much. The compilation was successful!
